numba/targets/arraymath.py

```diff
@@ -3541,6 +3542,32 @@ def np_extract_impl(condition, arr):
 
     return np_extract_impl
 
+  
+@overload(np.select)
+def np_select(condlist, choicelist, default=0):
+
+    def np_select_impl(condlist, choicelist, default=0):
+        out = default * np.ones(choicelist[0].shape, choicelist[0].dtype)
+        #should use reversed+zip, but reversed is not available
+        for i in range(len(condlist) - 1, -1, -1):
+            cond = condlist[i]
+            choice = choicelist[i]
+            out = np.where(cond, choice, out)
+        return out
+
+    if ((isinstance(condlist, (types.List, types.UniTuple)))
+        and (isinstance(choicelist, (types.List, types.UniTuple)))
+        and isinstance(condlist[0], types.Array)
+        and isinstance(condlist[0].dtype, types.Boolean)
+        and isinstance(default, (int, types.scalars.Number))):
+        if condlist[0].ndim != choicelist[0].ndim:
+            raise TypeError('condlist and choicelist elements must have the same number of dimensions')
+
+        if condlist[0].ndim < 1:
+            raise TypeError('condlist and choicelist elements must be arrays of at least dimension 1')
+        else:
```

Comment:
Seems like the number of entries in `condlist` and `choicelist` need to be the same, e.g.:
```python
def foo():
    x = np.arange(10)
    a = [x > 9, x > 8, x > 7, x > 6]
    b = [x, x**2, x,] # <--- oops one less
    return np.select(a, b, default=0)
```
NumPy raises an exception, whereas the current impl has an `IndexError`.

Reply:
thanks, I missed that check!

Reply:
@stuartarchibald , testing this case is throwing an error that I don't understand. I see that there's a pending deprecation warning. That's ok. But where is this coming from?
```
0 != 11

Expected :11
Actual   :0
```

Full text:

```
/home/.../numba/numba/tests/test_np_functions.py:197: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'choicelist' of function 'np_select.<locals>.np_select_arr_impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "targets/arraymath.py", line 3551:

    def np_select_arr_impl(condlist, choicelist, default=0):
    ^

  return np.select(condlist, choicelist, default=default)
/home/.../numba/numba/tests/test_np_functions.py:197: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'condlist' of function 'np_select.<locals>.np_select_arr_impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "targets/arraymath.py", line 3551:

    def np_select_arr_impl(condlist, choicelist, default=0):
    ^

  return np.select(condlist, choicelist, default=default)
/home/.../numba/numba/ir_utils.py:1959: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'choicelist' of function 'np_select'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "tests/test_np_functions.py", line 196:

def np_select(condlist, choicelist, default=0):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/.../numba/ir_utils.py:1959: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'condlist' of function 'np_select'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "tests/test_np_functions.py", line 196:

def np_select(condlist, choicelist, default=0):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))


0 != 11

Expected :11
Actual   :0
 <Click to see difference>

Traceback (most recent call last):
  File "/home/.../pycharm-2018.3.2/helpers/pycharm/teamcity/diff_tools.py", line 32, in _patched_equals
    old(self, first, second, msg)
  File "/home/.../anaconda3/envs/numbaenv/lib/python3.6/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/home/.../anaconda3/envs/numbaenv/lib/python3.6/unittest/case.py", line 822, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 11 != 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/.../anaconda3/envs/numbaenv/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/home/.../anaconda3/envs/numbaenv/lib/python3.6/unittest/case.py", line 608, in run
    self.tearDown()
  File "/home/.../mypyprojects/numba/numba/tests/support.py", line 651, in tearDown
    self.memory_leak_teardown()
  File "/home/.../mypyprojects/numba/numba/tests/support.py", line 625, in memory_leak_teardown
    self.assert_no_memory_leak()
  File "/home/.../mypyprojects/numba/numba/tests/support.py", line 634, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)



Ran 1 test in 0.711s

FAILED (failures=1)
```

Reply:
The `0 != 11` is from checking that the code is not leaking memory, and it is, because exceptions leak memory (it's a known issue). See the 6th bullet point here http://numba.pydata.org/numba-doc/dev/extending/overloading-guide.html#implementing-overload-for-numpy-functions  and this doc in general, it's new and intended to help contributors that are predominantly adding functionality through the use of `@overload`. Feedback on the doc is most welcome!

Reply:
I didn't know about that document, it's very good!